File: resetgo/internal/printer/printer.go
// Package printer implements the high-level printer operations (status, EEPROM
// read/write, waste counter reset) on top of a control backend, ported from the
// Python ez-reset tool.
package printer

import (
	"encoding/binary"
	"encoding/hex"
	"fmt"
	"strings"

	"ezreset/internal/control"
	"ezreset/internal/devices"
	"ezreset/internal/status"
	"ezreset/internal/utils"
)

// Printer wraps a control backend and a device definition.
type Printer struct {
	Device  devices.Device
	control control.ControlBackend
}

// New creates a Printer.
func New(backend control.ControlBackend, device devices.Device) *Printer {
	return &Printer{Device: device, control: backend}
}

func (p *Printer) sendCommand(command, payload []byte) ([]byte, error) {
	full := append(command, byte(len(payload)&0xFF), byte((len(payload)>>8)&0xFF))
	full = append(full, payload...)
	return p.control.Send(full)
}

func (p *Printer) sendFactoryCommand(model []byte, action int, payload []byte) ([]byte, error) {
	command := []byte("||")
	actionCode := []byte{
		byte(action),
		byte(action ^ 0xFF),
		byte(((action >> 1) & 0x7F) | ((action << 7) & 0x80)),
	}
	full := append(model, actionCode...)
	full = append(full, payload...)
	return p.sendCommand(command, full)
}

// GetStatus returns the parsed printer status.
func (p *Printer) GetStatus() (*status.Status, error) {
	expected := []byte("@BDC ST2\r\n")
	response, err := p.sendCommand([]byte("st"), []byte{0x01})
	if err != nil {
		return nil, err
	}
	if !strings.HasPrefix(string(response), string(expected)) {
		return nil, fmt.Errorf("unknown response %q for command %q", response, expected)
	}
	payload := response[len(expected):]
	return status.StatusFromBytes(payload)
}

func (p *Printer) readEEPROM(address int) (int, error) {
	action := 0x41
	expected := []byte("@BDC PS\r\n")
	response, err := p.sendFactoryCommand(p.Device.Model, action, uint16Bytes(address))
	if err != nil {
		return 0, err
	}
	if !strings.HasPrefix(string(response), string(expected)) {
		return 0, fmt.Errorf("unknown response %q for command %q", response, expected)
	}
	return parseIntHex(string(response[16:18])), nil
}

func (p *Printer) writeEEPROM(address, value int) ([]byte, error) {
	action := 0x42
	payload := append(uint16Bytes(address), byte(value))
	payload = append(payload, p.Device.Key...)
	return p.sendFactoryCommand(p.Device.Model, action, payload)
}

func (p *Printer) readEEPROMMultiple(addresses []int) ([]byte, error) {
	out := make([]byte, 0, len(addresses))
	for _, a := range addresses {
		v, err := p.readEEPROM(a)
		if err != nil {
			return nil, err
		}
		out = append(out, byte(v))
	}
	return out, nil
}

func (p *Printer) readEEPROMRange(address, size int) ([]byte, error) {
	action := 0x51
	expected := []byte("@BDC PS\r\n")
	payload := append(uint16Bytes(address), byte(size))
	response, err := p.sendFactoryCommand(p.Device.Model, action, payload)
	if err != nil {
		return nil, err
	}
	if !strings.HasPrefix(string(response), string(expected)) {
		return nil, fmt.Errorf("unknown response %q for command %q", response, expected)
	}
	return hex.DecodeString(strings.TrimSpace(string(response[16 : 16+size*2])))
}

// Identify returns the parsed IEEE 1284.4 identifier fields.
func (p *Printer) Identify() (map[string]string, error) {
	id, err := p.control.Identify()
	if err != nil {
		return nil, err
	}
	return utils.ParseIdentifier(id), nil
}

// GetSerial returns the printer serial number.
func (p *Printer) GetSerial() (string, error) {
	st, err := p.GetStatus()
	if err != nil {
		return "", err
	}
	return st.Serial, nil
}

// GetWaste returns the current waste counter values and their maxima.
func (p *Printer) GetWaste() ([][2]int, error) {
	var result [][2]int
	for _, counter := range p.Device.Counters {
		value, err := p.readEEPROMMultiple(counter.Addresses)
		if err != nil {
			return nil, err
		}
		result = append(result, [2]int{int(binary.LittleEndian.Uint16(value)), counter.Max})
	}
	return result, nil
}

// ResetWaste writes the reset values for all waste counters.
func (p *Printer) ResetWaste() error {
	for addr, value := range p.Device.Reset {
		if _, err := p.writeEEPROM(addr, value); err != nil {
			return err
		}
	}
	return nil
}

// Clean triggers a cleaning cycle at the given level.
func (p *Printer) Clean(level int) error {
	_, err := p.sendFactoryCommand(p.Device.Model, 0x84, []byte{byte(level)})
	return err
}

// PowerOff powers the printer off.
func (p *Printer) PowerOff() error {
	_, err := p.sendFactoryCommand(p.Device.Model, 0x20, nil)
	return err
}

// Restart restarts the printer.
func (p *Printer) Restart() error {
	_, err := p.sendFactoryCommand(p.Device.Model, 0x21, nil)
	return err
}

func uint16Bytes(v int) []byte {
	return []byte{byte(v & 0xFF), byte((v >> 8) & 0xFF)}
}

func parseIntHex(s string) int {
	v, _ := hex.DecodeString(strings.TrimSpace(s))
	if len(v) == 0 {
		return 0
	}
	return int(v[0])
}

File: resetcplusplus/src/printer.cpp
#include "printer.h"
#include "utils.h"

#include <stdexcept>
#include <string>

namespace ezreset {

std::vector<unsigned char> Printer::sendCommand(const std::vector<unsigned char>& command,
                                                const std::vector<unsigned char>& payload) {
    std::vector<unsigned char> full = command;
    full.push_back(static_cast<unsigned char>(payload.size() & 0xFF));
    full.push_back(static_cast<unsigned char>((payload.size() >> 8) & 0xFF));
    full.insert(full.end(), payload.begin(), payload.end());
    return backend_->send(full);
}

std::vector<unsigned char> Printer::sendFactoryCommand(const std::vector<unsigned char>& model,
                                                       int action,
                                                       const std::vector<unsigned char>& payload) {
    std::vector<unsigned char> command = {'|', '|'};
    std::vector<unsigned char> actionCode = {
        static_cast<unsigned char>(action),
        static_cast<unsigned char>(action ^ 0xFF),
        static_cast<unsigned char>(((action >> 1) & 0x7F) | ((action << 7) & 0x80))};
    std::vector<unsigned char> full = model;
    full.insert(full.end(), actionCode.begin(), actionCode.end());
    full.insert(full.end(), payload.begin(), payload.end());
    return sendCommand(command, full);
}

Status Printer::getStatus() {
    std::vector<unsigned char> expected = {'@','B','D','C',' ','S','T','2','\r','\n'};
    auto response = sendCommand({'s','t'}, {'\x01'});
    if (response.size() < expected.size()
        || std::string(response.begin(), response.begin() + expected.size())
               != std::string(expected.begin(), expected.end())) {
        throw std::runtime_error("unknown status response");
    }
    std::vector<unsigned char> payload(response.begin() + expected.size(), response.end());
    return statusFromBytes(payload);
}

int Printer::readEEPROM(int address) {
    int action = 0x41;
    std::vector<unsigned char> expected = {'@','B','D','C',' ','P','S','\r','\n'};
    std::vector<unsigned char> addr = {
        static_cast<unsigned char>(address & 0xFF),
        static_cast<unsigned char>((address >> 8) & 0xFF)};
    auto response = sendFactoryCommand(device_.model, action, addr);
    if (response.size() < expected.size()
        || std::string(response.begin(), response.begin() + expected.size())
               != std::string(expected.begin(), expected.end())) {
        throw std::runtime_error("unknown EEPROM response");
    }
    std::string hex(response.begin() + 16, response.begin() + 18);
    return static_cast<int>(std::stoul(hex, nullptr, 16));
}

std::vector<unsigned char> Printer::readEEPROMMultiple(const std::vector<int>& addresses) {
    std::vector<unsigned char> out;
    for (int a : addresses) out.push_back(static_cast<unsigned char>(readEEPROM(a)));
    return out;
}

std::vector<unsigned char> Printer::readEEPROMRange(int address, int size) {
    int action = 0x51;
    std::vector<unsigned char> expected = {'@','B','D','C',' ','P','S','\r','\n'};
    std::vector<unsigned char> payload = {
        static_cast<unsigned char>(address & 0xFF),
        static_cast<unsigned char>((address >> 8) & 0xFF),
        static_cast<unsigned char>(size & 0xFF)};
    auto response = sendFactoryCommand(device_.model, action, payload);
    if (response.size() < expected.size()
        || std::string(response.begin(), response.begin() + expected.size())
               != std::string(expected.begin(), expected.end())) {
        throw std::runtime_error("unknown EEPROM range response");
    }
    std::string hex(response.begin() + 16, response.begin() + 16 + size * 2);
    std::vector<unsigned char> out;
    for (size_t i = 0; i + 1 < hex.size(); i += 2) {
        out.push_back(static_cast<unsigned char>(std::stoul(hex.substr(i, 2), nullptr, 16)));
    }
    return out;
}

void Printer::writeEEPROM(int address, int value) {
    int action = 0x42;
    std::vector<unsigned char> payload = {
        static_cast<unsigned char>(address & 0xFF),
        static_cast<unsigned char>((address >> 8) & 0xFF),
        static_cast<unsigned char>(value & 0xFF)};
    payload.insert(payload.end(), device_.key.begin(), device_.key.end());
    sendFactoryCommand(device_.model, action, payload);
}

std::vector<std::pair<int, int>> Printer::getWaste() {
    std::vector<std::pair<int, int>> result;
    for (const auto& counter : device_.counters) {
        auto value = readEEPROMMultiple(counter.addresses);
        int v = 0;
        for (size_t i = 0; i < value.size(); i++) {
            v |= (static_cast<int>(value[i]) << (8 * i));
        }
        result.push_back({v, counter.max});
    }
    return result;
}

void Printer::resetWaste() {
    for (const auto& kv : device_.reset) {
        writeEEPROM(kv.first, kv.second);
    }
}

std::string Printer::getSerial() {
    return getStatus().serial;
}

void Printer::clean(int level) {
    sendFactoryCommand(device_.model, 0x84, {static_cast<unsigned char>(level & 0xFF)});
}

void Printer::powerOff() {
    sendFactoryCommand(device_.model, 0x20, {});
}

void Printer::restart() {
    sendFactoryCommand(device_.model, 0x21, {});
}

} // namespace ezreset
